File: ImperatorToCK3/Source/CK3/Titles/TitlesHistory.cpp
#include "TitlesHistory.h"
#include "CommonUtilities/HistoryFactory.h"
#include "Configuration/Configuration.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"
#include "OSCompatibilityLayer.h"



CK3::TitlesHistory::TitlesHistory(const std::string& folderPath) {
	auto filenames = commonItems::GetAllFilesInFolderRecursive(folderPath);
	LOG(LogLevel::Info) << "-> Parsing title history.";
	registerKeys();
	for (const auto& fileName : filenames) {
		parseFile(folderPath + "/" + fileName);
	}
	clearRegisteredKeywords();
	LOG(LogLevel::Info) << "<> Loaded " << historyMap.size() << " title histories.";
}

std::optional<CK3::TitleHistory> CK3::TitlesHistory::popTitleHistory(const std::string& titleName) {
	if (const auto historyItr = historyMap.find(titleName); historyItr != historyMap.end()) {
		auto history = historyItr->second;
		historyMap.erase(titleName);
		return history;
	}
	return std::nullopt;
}


void CK3::TitlesHistory::TitlesHistory::registerKeys() {
	registerRegex(R"((e|k|d|c|b)_[A-Za-z0-9_\-\']+)", [this](const std::string& titleName, std::istream& theStream) {
		const auto historyItem = commonItems::stringOfItem(theStream).getString();
		std::stringstream tempStream(historyItem);
		if (historyItem.find('{') != std::string::npos) {
			const auto history = historyFactory.getHistory(tempStream);
			historyMap.emplace(titleName, TitleHistory(history));
		}
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

File: ImperatorToCK3/CK3/Titles/TitlesHistory.cs
﻿using commonItems;
using commonItems.Collections;
using ImperatorToCK3.CommonUtils;
using System.Collections.Generic;

namespace ImperatorToCK3.CK3.Titles;

public class TitlesHistory : Parser {
	public TitlesHistory() { }
	public TitlesHistory(string folderPath) {
		var filenames = SystemUtils.GetAllFilesInFolderRecursive(folderPath);
		Logger.Info("Parsing title history...");
		RegisterKeys();
		foreach (var filename in filenames) {
			ParseFile(System.IO.Path.Combine(folderPath, filename));
		}
		ClearRegisteredRules();
		Logger.Info($"Loaded {historyDict.Count} title histories.");
	}
	public TitleHistory? PopTitleHistory(string titleName) { // "pop" as from stack, not Imperator Pop ;)
		if (!historyDict.TryGetValue(titleName, out var historyToReturn)) {
			return null;
		}
		historyDict.Remove(titleName);
		return historyToReturn;
	}

	private void RegisterKeys() {
		RegisterRegex(@"(e|k|d|c|b)_[A-Za-z0-9_\-\']+", (reader, titleName) => {
			var historyItem = new StringOfItem(reader).ToString();
			if (!historyItem.Contains('{')) {
				return;
			}
			var tempReader = new BufferedReader(historyItem);
			if (historyDict.TryGetValue(titleName, out var existingHistory)) {
				existingHistory.Update(historyFactory, tempReader);
			} else {
				var history = historyFactory.GetHistory(tempReader);
				historyDict.Add(titleName, new TitleHistory(history));
			}
		});
		RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
	}

	private readonly HistoryFactory historyFactory = new HistoryFactory.HistoryFactoryBuilder()
		.WithSimpleField("holder", new OrderedSet<string>{"holder", "holder_ignore_head_of_faith_requirement"}, null)
		.WithSimpleField("government", "government", null)
		.WithSimpleField("liege", "liege", null)
		.WithSimpleField("development_level", "change_development_level", null)
		.WithSimpleField("succession_laws", "succession_laws", new SortedSet<string>())
		.Build();
	private readonly Dictionary<string, TitleHistory> historyDict = new();
}